### ISSUE TITLE
[Bug]: `react/prefer-read-only-props` doesn't work with implicit `React` reference (`react/jsx-runtime`)

### ISSUE THREAD

**Issue by alexilyaev** [jsx-eslint/eslint-plugin-react#3650]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

Adding `react/prefer-read-only-props` rule does not trigger when using `React.FC` without importing `React`.

In modern React, we don't need to import `React` thanks to the new JSX Transform:
- https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#whats-different-in-the-new-transform
And TypeScript supports this out of the box:
- https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#jsx-factories

So even for React types we don't have to import them. `React.FC` just works in `.tsx` files.

But `react/prefer-read-only-props` fails to identify the component as a React component if we use `React.FC` without importing `React`.

### `.eslintrc.js`

```js
module.exports = {
  root: true,
  env: {
    browser: true,
    es6: true,
    es2021: true,
  },
  plugins: ['@typescript-eslint', 'react'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2021,
    sourceType: 'module',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  rules: {
    'react/prefer-read-only-props': 'warn',
  },
};
```
### `tsconfig.json`
```json
{
  "compilerOptions": {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "baseUrl": "./",
    "composite": false,
    "declaration": false,
    "declarationMap": false,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "module": "ESNext",
    "moduleResolution": "node",
    "noEmit": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "target": "ES2021",
    "types": ["node", "@testing-library/jest-dom"],
  },
  "include": ["*.ts", "src/**/*.ts", "src/**/*.tsx", "src/**/*.json"],
  "exclude": ["node_modules", "build", "storybook-static"]
}

```
Tried in VS Code and in the command line:
```shell
yarn eslint --max-warnings 0 src/components/Chip.tsx
```

## This code does NOT error
```tsx
interface ChipProps {
  chipColor: string;
  label: string;
}

const Chip: React.FC<ChipProps> = ({ chipColor, label }) => {
  return <div style={{ backgroundColor: chipColor }}>{label}</div>;
};
```
<img width="685" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/f6392cd2-1ae4-465d-b745-7a12b76f2a37">

## This code does
import React from 'react'

```tsx
interface ChipProps {
  chipColor: string;
  label: string;
}

const Chip: React.FC<ChipProps> = ({ chipColor, label }) => {
  return <div style={{ backgroundColor: chipColor }}>{label}</div>;
};
```

<img width="686" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/fb4f3c16-bfd2-4545-914f-52713d7dca29">

### eslint-plugin-react version

7.33.2

### eslint version

8.38.0

### node version

v16.14.0

**Comment by ljharb** [comment]
Oof, that seems really unfortunate from TS - referencing React without importing it is horrific. The new jsx transform works because it doesn’t transform to something that references React.

I’m not sure how we can handle this cleanly.

**Comment by alexilyaev** [comment]
@ljharb Can you point me to the source logic that recognizes `React.FC` with the `import React` but that fails without the `import`?
Is this rule "type-aware"? Meaning can it read the types?

**Comment by ljharb** [comment]
It's in lib/util/propTypes.js, and yes, it's type-aware whenever the parser provides that information.

The key is, we shouldn't be treating something as a React type unless we're certain it came from React.

**Comment by alexilyaev** [comment]
> The key is, we shouldn't be treating something as a React type unless we're certain it came from React
Even if the identifier is `React.FC`?

**Comment by alexilyaev** [comment]
Btw, the rule already identifies a React component even without any import matching:
```tsx
interface ChipProps {
  chipColor: string;
  label: string;
}

const Chip = ({ chipColor, label }: ChipProps) => {
  return <div style={{ backgroundColor: chipColor }}>{label}</div>;
};
```
![image](https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/11fe44c6-2593-4452-916c-c0d97a2e6b0c)

Can't we use the same heuristic to assume that when we see `React.FC` (without importing React) AND the function returns JSX (or however it is determined) to mean it is indeed a React component?

**Comment by ljharb** [comment]
Yes, we can never assume any identifier is anything specific; someone could make their own thing called that.

a React component uses jsx syntax, and `jsx-uses-react` ensures that jsx syntax imports React (when not using jsx-runtime).

If it's TS that provides React.FC as an implicit global, then certainly we could look for when `React.FC` is used in type space, and is also NOT imported or otherwise registered as a global to eslint?

**Comment by alexilyaev** [comment]
Indeed `React` in this pattern is provided only in type space.
This works because React with TypeScript projects install the `@types/react-dom` dependency (which also installs `@types/react`), since without these, there's a whole bunch of type errors.

<img width="1517" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/96db1043-2f52-4d76-b7bb-704205edb9e0">
<img width="1688" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/75a01fe3-88bb-4106-8731-c9ba4e7aba23">
<br /><br />

Considering this fact, I think it's harder to imagine someone using `React.FC` that is not the actual React type.
What do you think?

### Side note

This issue is also relevant for `react/prop-types` rule, which has the same heuristic in order to disable itself:
https://github.com/vedadeepta/eslint-plugin-react/blob/master/lib/util/propTypes.js#L516 -> `isValidReactGenericTypeAnnotation`

And currently fails to do so with the mentioned pattern:
<img width="1045" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/480469/b6d7f81e-0b88-4fef-8db6-ee2ed9d95f8f">

**Comment by ljharb** [comment]
I'd still like to be precise about ensuring that `React.FC`, when React is not available, is only recognized as such when it's in type-space and React is not a global.
